src/transformers/models/regnet/modeling_tf_regnet.py

```diff
@@ -0,0 +1,581 @@
+
+
+# Copied from:
+# https://gist.github.com/Rocketknight1/43abbe6e73f1008e6e459486e01e0ceb
+class TFAdaptiveAvgPool1D(tf.keras.layers.Layer):
+    def __init__(self, output_dim, mode="dense", **kwargs):
+        super().__init__(**kwargs)
+        self.output_dim = output_dim
+        self.mode = mode
+        self.map = None
+
+    def build(self, input_shape):
+        super().build(input_shape)
+        """We pre-compute the sparse matrix for the build() step once. The below code comes
+        from https://stackoverflow.com/questions/53841509/how-does-adaptive-pooling-in-pytorch-work/63603993#63603993."""
+
+        def get_kernels(ind, outd) -> List:
+            """Returns a List [(kernel_offset_start,kernel_length)] defining all the pooling kernels for a 1-D adaptive
+            pooling layer that takes an input of dimension `ind` and yields an output of dimension `outd`"""
+
+            def start_index(a, b, c):
+                return math.floor((float(a) * float(c)) / b)
+
+            def end_index(a, b, c):
+                return math.ceil((float(a + 1) * float(c)) / b)
+
+            results = []
+            for ow in range(outd):
+                start = start_index(ow, outd, ind)
+                end = end_index(ow, outd, ind)
+                sz = end - start
+                results.append((start, sz))
+            return results
+
+        in_dim = int(input_shape[-1])
+        kernels = get_kernels(in_dim, self.output_dim)
+        sparse_map = np.zeros((in_dim, self.output_dim), dtype=np.float32)
+        for i, kernel in enumerate(kernels):
+            sparse_map[kernel[0] : kernel[0] + kernel[1], i] = 1 / kernel[1]
+        if self.mode == "dense":
+            self.map = tf.constant(sparse_map)
+        else:
+            self.map = tf.sparse.from_dense(sparse_map)
+
+    def call(self, inputs):
+        if self.mode == "dense":
+            return inputs @ self.map
+        else:
+            input_dims = inputs.shape
+            input_matrix = tf.reshape(inputs, (-1, input_dims[-1]))
+            out = tf.sparse.sparse_dense_matmul(input_matrix, self.map)
+            return tf.reshape(out, input_dims[:-1].as_list() + [-1])
+
+
+class TFAdaptiveAvgPool2D(tf.keras.layers.Layer):
+    def __init__(self, output_shape, mode="dense", **kwargs):
+        super().__init__(**kwargs)
+        self.h_pool = TFAdaptiveAvgPool1D(output_shape[0], mode=mode)
+        self.w_pool = TFAdaptiveAvgPool1D(output_shape[1], mode=mode)
+
+    def call(self, inputs):
+        # Rearrange from NHWC -> NCHW
+        inputs = tf.transpose(inputs, perm=[0, 3, 1, 2])
+        # Perform W-pooling
+        inputs = self.w_pool(inputs)
+        # Rearrange NCHW -> NCWH
+        inputs = tf.transpose(inputs, perm=[0, 1, 3, 2])
+        # Perform H-pooling
+        inputs = self.h_pool(inputs)
+        # Rearrange from NCWH -> NHWC
+        inputs = tf.transpose(inputs, perm=[0, 3, 2, 1])
```

Comment:
It seems a bit inefficient to rearrange the channels one way and then the other. Can't we just pool on dimensions 2 and 3 instead?

Reply:
@Rocketknight1

Reply:
I couldn't find a way to do it given how we have to implement this pool, unfortunately. The problem is that Torch's `AdaptivePool` computes the positions to pool together in its own singular way that TF doesn't have an op for, so I effectively have to "emulate" it with two inefficient matmuls instead. I tried two other solutions, firstly merging WH and then doing a dense-sparse matmul, and secondly using depthwise convolutions with different kernels followed by a selection step, but this approach was the fastest of the three.

Possibly I could use an `einsum` instead of a matmul and avoid the rearrangements, but I was running out of time at this point! If we decide it's critical later I can take another pass at it, or in the worst case port the op to TF in CUDA and submit it to TF-addons, and then we can use that as a dependency.

Reply:
I'm sure you've probably already seen, but just in case you haven't, tensorflow add-ons have adaptive average pooling layers e.g. [AdaptiveAveragePooling2D](https://github.com/tensorflow/addons/blob/8cec33fcaaf1cf90aec7bdd55a0fcdbb251ce5c2/tensorflow_addons/layers/adaptive_pooling.py#L157-L234). 

I don't know the gritty implementation details for this or implication for this model, but I was able to get the same outputs as the equivalent pytorch layer when adapting these for Swin.

Reply:
I saw those, but the TFA `AdaptiveAveragePooling` layers are only equivalent to the PyTorch layer when the output shape evenly divides the input shape! When the output shape doesn't evenly divide the input shape, then PyTorch `AdaptivePool` scatters potentially-overlapping pool windows across the input with different widths, using its own idiosyncratic algorithm to decide where to put them. It is impossible to replicate that with any pooling operation that has a uniform width (i.e. all of the TF pooling operations). You need to use some kind of multiplication/convolution operation that lets you mask positions with 0 values, which is what I did here.

If `RegNet` always only resizes by clean integer factors (I don't think it does, but it's been a while since I looked - @sayakpaul can you confirm?) then we could use the TFA op, but it's definitely not a correct and general reimplementation of the PyTorch op outside of that special case.

Reply:
There's a good discussion of the (mostly undocumented) internals of that PyTorch op [here](https://stackoverflow.com/questions/53841509/how-does-adaptive-pooling-in-pytorch-work/63603993#63603993).

Reply:
> If RegNet always only resizes by clean integer factors (I don't think it does, but it's been a while since I looked - @sayakpaul can you confirm?)

I can confirm.